ansible_runner/streaming.py

```diff
@@ -0,0 +1,221 @@
+    'forks',
+    'host_pattern',
+    'ident',
+    'ignore_logging',
+    'inventory',
+    'limit',
+    'module',
+    'module_args',
+    'omit_event_data',
+    'only_failed_event_data',
+    'playbook',
+    'verbosity',
+)
+
+
+class Transmitter(object):
+    def __init__(self, _output=None, **kwargs):
+        if _output is None:
+            _output = sys.stdout.buffer
+        self._output = _output
+        self.kwargs = kwargs
+        self.config = ansible_runner.RunnerConfig(**kwargs)
+
+        self.status = "unstarted"
+        self.rc = None
+
+    def run(self):
+        self.config.prepare()
+        remote_options = {key: value for key, value in self.kwargs.items() if key in remote_run_options}
+
+        private_data_dir = self.kwargs.get('private_data_dir', None)
+        self._output.write(
+            utils.stream_dir(private_data_dir, kwargs=json.dumps(remote_options, cls=UUIDEncoder)))
+        self._output.write(json.dumps({'eof': True}).encode('utf-8'))
+        self._output.write(b'\n')
+        self._output.flush()
+        # self._output.close()
+
+        return self.status, self.rc
+
+
+class Worker(object):
+    def __init__(self, _input=None, _output=None, **kwargs):
+        if _input is None:
+            _input = sys.stdin.buffer
+        if _output is None:
+            _output = sys.stdout.buffer
+        self._input = _input
+        self._output = _output
+
+        self.kwargs = kwargs
+
+        private_data_dir = kwargs.get('private_data_dir')
+        if not private_data_dir:
+            private_data_dir = tempfile.TemporaryDirectory().name
+        self.private_data_dir = private_data_dir
+
+        self.status = "unstarted"
+        self.rc = None
+
+    def run(self):
+        while True:
+            line = self._input.readline()
+            data = json.loads(line)
+
+            if 'zipfile' in data:
+                zip_data = self._input.read(data['zipfile'])
+                buf = io.BytesIO(zip_data)
+                with zipfile.ZipFile(buf, 'r') as archive:
+                    archive.extractall(path=self.private_data_dir)
+            elif 'eof' in data:
+                break
+
+        kwargs_path = os.path.join(self.private_data_dir, 'kwargs')
+        if os.path.exists(kwargs_path):
+            with open(kwargs_path, "r") as kwf:
+                kwargs = json.load(kwf)
+            if not isinstance(kwargs, dict):
+                raise ValueError("Invalid kwargs data")
+        else:
+            kwargs = {}
+
+        self.kwargs.update(kwargs)
+
+        self.kwargs['quiet'] = True
+        self.kwargs['private_data_dir'] = self.private_data_dir
+        self.kwargs['status_handler'] = self.status_handler
+        self.kwargs['event_handler'] = self.event_handler
+        self.kwargs['artifacts_handler'] = self.artifacts_handler
+        self.kwargs['finished_callback'] = self.finished_callback
+
+        r = ansible_runner.interface.run(**self.kwargs)
```

Comment:
This calls `init_runner` which creates a `RunnerConfig` and calls prepare. But this is `Worker`, and the `Transmitter` already did those things by the time it got here.

I want to get some hard and fast rules about separation of responsibilities. As both the transmitter and worker are creating runner configs, there are going to be bugs, because the two are going to be diverge somehow or another.

Reply:
Agreed, which is why I'm sitting down and figuring out where each `RunnerConfig` attribute is coming from, and whose responsibility it should be.

Reply:
Potentially unpopular opinion: `Transmitter` should not be responsible for creating a `RunnerConfig` at all.

Reply:
The downside is that validation that the command can be constructed in a meaningful way happens late, in `Worker`.

Reply:
You mean validation of the command line parameters?  The whole transmit/worker/process scheme should only be called by applications like Tower that are written not to issue wrong commands.

In the unusual case that a user is running this directly and types a command that doesn't make sense, but makes enough sense to get transmitted to a worker, I think it's fine if it's the worker that detects the problem.